polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,291 @@
+        )
+        for row in cursor.fetchall():
+            uprn = row[0]
+            polling_station_id = row[1]
+            if polling_station_id not in polling_station_lookup:
+                polling_station_lookup[polling_station_id] = {uprn}
+            else:
+                polling_station_lookup[polling_station_id].add(uprn)
+
+        return polling_station_lookup
+
+
+class StationSet(CustomSet, AssignPollingStationsMixin):
+    def build_namedtuple(self, element):
+
+        # Point is mutable, so we must serialize it to store in a tuple
+        if "location" in element and element["location"]:
+            location = element["location"].ewkb  # use ewkb so it encodes srid
+        else:
+            location = None
+
+        return Station(
+            element["council"],
+            element["internal_council_id"],
+            element.get("postcode", ""),
+            element.get("address", ""),
+            location,
+            element.get("polling_district_id", ""),
+        )
+
+    def get_polling_station_lookup(self):
+        """
+        for each address, build a lookup of polling_station_id -> set of uprns
+        This does the lookup in the db so you need to have called save() before
+        using this.
+        """
+        polling_station_lookup = {}
+        cursor = connection.cursor()
+        cursor.execute(
+            """ 
+                SELECT a.uprn, s.internal_council_id
+                FROM addressbase_address a
+                    JOIN addressbase_uprntocouncil u
+                    ON a.uprn = u.uprn
+                    JOIN pollingstations_pollingdistrict d
+                    ON ST_Contains(d.area, a.location)
+                    JOIN pollingstations_pollingstation s
+                    ON s.polling_district_id = d.internal_council_id
+                WHERE d.council_id=%s
+                AND u.lad=%s
+                AND s.council_id=%s
+            """,
+            [self.council_id, self.council_id, self.council_id],
+        )
+        for row in cursor.fetchall():
+            uprn = row[0]
+            polling_station_id = row[1]
+            if polling_station_id not in polling_station_lookup:
+                polling_station_lookup[polling_station_id] = {uprn}
+            else:
+                polling_station_lookup[polling_station_id].add(uprn)
+
+        return polling_station_lookup
+
+    def save(self):
+        stations_db = []
+        for station in self.elements:
+            record = PollingStation(
+                council=station.council,
+                internal_council_id=station.internal_council_id,
+                postcode=station.postcode,
+                address=station.address,
+                location=station.location,
+                polling_district_id=station.polling_district_id,
+            )
+            stations_db.append(record)
+        PollingStation.objects.bulk_create(stations_db)
+
+
+class AddressList(AssignPollingStationsMixin):
```

Comment:
Wow. This really removes a lot of the complexity in here :wave:
I wonder how long it can stay that way